kernel/src/hil/symmetric_encryption.rs

```diff
@@ -1,42 +1,88 @@
-    fn set_key(&self, key: &'static mut [u8], len: usize) -> &'static mut [u8];
+    /// Set the IV (or initial counter); returns `EINVAL` if length is not `AES128_BLOCK_SIZE`
+    fn set_iv(&self, iv: &[u8]) -> ReturnCode;
 
-    /// encryption and decryption for aes in counter mode
-    /// because only the encryption-mode of the cipher only one method is needed
-    /// other chips perhaps only ignore "init_ctr" and assume all is performed in HW
-    fn aes128_crypt_ctr(&self, data: &'static mut [u8], init_ctr: &'static mut [u8], len: usize);
+    /// Set the source buffer.  If this is full, the encryption
+    /// input will be this entire buffer, and its size must match
+    /// `stop_index - start_index` when `crypt()` is called.
+    /// If this is empty, the destination buffer will be read
+    /// to provide the plaintext input.
+    fn set_source(&'a self, buf: Option<&'a mut [u8]>) -> ReturnCode;
+
+    /// Return the source buffer, if any
+    fn take_source(&'a self) -> Option<&'a mut [u8]>;
+
+    /// Set the destination buffer.  If `set_source()` has not
+    /// been used to pass a source buffer, this buffer will also
+    /// provide the encryption input, which will be overwritten.
+    /// The option should be full whenever `crypt()` is called.
+    /// Returns SUCCESS if the buffer was installed, or EBUSY
+    /// if the encryption unit is still busy.
+    fn put_dest(&'a self, dest: Option<&'a mut [u8]>) -> ReturnCode;
+
+    /// Return the destination buffer, if any.
+    /// Returns EBUSY if the encryption unit is still busy.
+    fn take_dest(&'a self) -> Result<Option<&'a mut [u8]>, ReturnCode>;
+
+    /// Begin a new message (with the configured IV) when `crypt()` is next
+    /// called.  Multiple calls to `crypt()` (accompanied by `set_source()` or
+    /// `put_dest`) may be made between calls to `start_message()`, allowing the
+    /// encryption context to extend over non-contiguous extents of data.
+    fn start_message(&self);
+
+    /// Request an encryption/decryption
+    ///
+    /// The indices `start_index` and `stop_index` must be valid offsets in
+    /// a buffer previously passed in with `set_data`, and the length
+    /// `stop_index - start_index` must be a multiple of
+    /// `AES128_BLOCK_SIZE`.  Otherwise, INVAL will be returned.
+    ///
+    /// If SUCCESS is returned, the client's `crypt_done` method will eventually
+    /// be called, and the portion of the data buffer between `start_index`
+    /// and `stop_index` will hold the result of the encryption/decryption.
+    ///
+    /// For correct operation, the methods `set_key` and `set_iv` must have
+    /// previously been called to set the buffers containing the
+    /// key and the IV (or initial counter value), and a method `set_mode_*()`
+    /// must have been called to set the desired mode.  These settings persist
+    /// across calls to `crypt()`.
+    fn crypt(&self, start_index: usize, stop_index: usize) -> ReturnCode;
```

Comment:
Hmm, what is the reason of using crypt in the AES trait over using specific methods in the AESCBC and AESCTR crate?

IMO, encrypt and decrypt makes sense for AESCBC and crypt makes sense for AESCTR